## Review: `refactor: centralize OS path normalization (#1156)`

### Verdict: ⚠️ Changes Requested — one test failure on Linux; otherwise excellent

---

### What Works Well ✅

The `NormalizePath` and `Normalize` replacements are semantically identical to the old code. All affected packages pass `-count=3 -race` on Linux:

| Package | Result |
|---------|--------|
| `internal/pkg/filepathutil` (all 10 subtests) | ✅ PASS |
| `internal/infrastructure/security` | ✅ PASS |
| `internal/tools/workspace` (path tests) | ✅ PASS |
| `internal/tools/analysis/analysistest` | ✅ PASS |
| `internal/tools/analysis` (harvest + deadcode) | ✅ PASS |

The new package is clean, well-documented, and the three-function API (`Normalize`, `NormalizePath`, `NormalizeKey`) covers all use cases.

---

### ❌ Blocking: `TestIsInSearchPath/Win_root` fails on Linux

```
--- FAIL: TestIsInSearchPath/Win_root (0.00s)
    index_search_test.go:70: isInSearchPath("C:/", "C:/a/b.go") = false; want true
```

**Root cause**: `NormalizeKey` calls `Normalize` → `EvalSymlinks`. The old `isInSearchPath` code only did `ToSlash` + `ToLower` — **no filesystem calls**. On Linux, `C:/` is not a real path, so `EvalSymlinks` fails and the recursive fallback misreconstructs it:

```
NormalizeKey("C:/")
 → Normalize("C:/")
   → EvalSymlinks("C:/") → ENOENT
   → Dir("C:/")  → "C:"     ← Unix treats C: as a directory name
   → Base("C:/") → "C:"     ← Same as Dir!
   → Join("C:", "C:") → "C:/C:"  ← doubled component
   → ToLower → "c:/c:"           ← ≠ expected "c:/"
```

On Linux, `filepath.Dir("C:/")` and `filepath.Base("C:/")` both return `"C:"` because the Unix `filepath` package treats `C:` as a regular directory name, not a volume prefix. The reconstruction `Join("C:", "C:")` produces `"C:/C:"` instead of `"C:/"`.

This is not just a test issue — `NormalizeKey` changes the semantics of `isInSearchPath` from pure string manipulation to filesystem-touching path resolution.

### Recommended Fix

Add a symlink-free variant for `isInSearchPath`:

```go
// NormalizeForCompare converts to forward slashes, strips volume prefix,
// and lowercases. Unlike NormalizeKey, it does NOT resolve symlinks —
// suitable when both paths come from the same OS and only format
// normalization is needed.
func NormalizeForCompare(path string) string {
    s := filepath.ToSlash(path)
    if vol := filepath.VolumeName(s); vol != "" {
        s = s[len(vol):]
    }
    return strings.ToLower(s)
}
```

Then in `index_search.go`:

```go
func (idx *indexer) isInSearchPath(targetPath, filePath string) bool {
    targetPath = filepathutil.NormalizeForCompare(targetPath)
    filePath   = filepathutil.NormalizeForCompare(filePath)
    // ... rest unchanged
```

This preserves the exact old behavior while still centralizing the logic. The other `NormalizeKey` call sites (`isInTargetScope`, `mock_security.IsPathSafe`) work with real on-disk paths where `EvalSymlinks` succeeds, so they are fine as-is.

---

### Minor Observation (non-blocking)

`security.pathPolicy.resolveSymlinks` is now a one-line wrapper around `filepathutil.NormalizePath`. Consider removing it entirely in a follow-up and calling `NormalizePath` directly at the two remaining call sites (`ValidatePath` and `checkBoundary`).

--- a/internal/pkg/filepathutil/normalize.go
+++ b/internal/pkg/filepathutil/normalize.go
@@ -70,8 +70,14 @@ func NormalizeKey(path string) string {
 // already absolute and on the same filesystem).
 func NormalizeForCompare(path string) string {
 	s := filepath.ToSlash(path)
+	// Strip Windows volume prefix on all platforms.
+	// filepath.VolumeName only detects volumes on the current OS,
+	// so we also check for [A-Za-z]: manually for cross-platform use.
 	if vol := filepath.VolumeName(s); vol != "" {
 		s = s[len(vol):]
+	} else if len(s) >= 2 && s[1] == ':' &&
+		((s[0] >= 'A' && s[0] <= 'Z') || (s[0] >= 'a' && s[0] <= 'z')) {
+		s = s[2:]
 	}
 	return strings.ToLower(s)
 }

--- a/internal/pkg/filepathutil/normalize.go
+++ b/internal/pkg/filepathutil/normalize.go
@@ -70,6 +70,10 @@ func NormalizeKey(path string) string {
 // already absolute and on the same filesystem).
 func NormalizeForCompare(path string) string {
 	s := filepath.ToSlash(path)
+	// On non-Windows platforms, ToSlash does not convert backslashes
+	// because \ is not the OS path separator. Normalize them manually
+	// so that Windows paths (e.g., D:\foo\bar) work cross-platform.
+	s = strings.ReplaceAll(s, "\\", "/")
 	// Strip Windows volume prefix on all platforms.
 	// filepath.VolumeName only detects volumes on the current OS,
 	// so we also check for [A-Za-z]: manually for cross-platform use.

--- a/internal/pkg/filepathutil/normalize.go
+++ b/internal/pkg/filepathutil/normalize.go
@@ -62,3 +62,16 @@ func NormalizeKey(path string) string {
 	}
 	return strings.ToLower(s)
 }
+
+// NormalizeForCompare converts to forward slashes, strips the Windows
+// volume prefix, and lowercases. Unlike NormalizeKey, it does NOT resolve
+// symlinks — suitable when both paths come from the same OS and only
+// format normalization is needed (e.g., comparing two paths that are
+// already absolute and on the same filesystem).
+func NormalizeForCompare(path string) string {
+	s := filepath.ToSlash(path)
+	if vol := filepath.VolumeName(s); vol != "" {
+		s = s[len(vol):]
+	}
+	return strings.ToLower(s)
+}

--- a/internal/pkg/filepathutil/normalize_test.go
+++ b/internal/pkg/filepathutil/normalize_test.go
@@ -139,3 +139,40 @@ func TestNormalize_Idempotent(t *testing.T) {
 		t.Errorf("Normalize not idempotent: %q vs %q", first, second)
 	}
 }
+
+func TestNormalizeForCompare_WindowsRoot(t *testing.T) {
+	// C:/ must normalize to / without touching the filesystem.
+	// Volume prefix is stripped, leaving just the root separator.
+	result := NormalizeForCompare("C:/")
+	if result != "/" {
+		t.Errorf("NormalizeForCompare(%q) = %q; want %q", "C:/", result, "/")
+	}
+}
+
+func TestNormalizeForCompare_StripsVolume(t *testing.T) {
+	result := NormalizeForCompare("D:\\Users\\foo\\bar.txt")
+	if result != "/users/foo/bar.txt" {
+		t.Errorf("NormalizeForCompare = %q; want %q", result, "/users/foo/bar.txt")
+	}
+}
+
+func TestNormalizeForCompare_WindowsChildPath(t *testing.T) {
+	result := NormalizeForCompare("C:\\a\\b.go")
+	if result != "/a/b.go" {
+		t.Errorf("NormalizeForCompare = %q; want %q", result, "/a/b.go")
+	}
+}
+
+func TestNormalizeForCompare_UnixPath(t *testing.T) {
+	result := NormalizeForCompare("/usr/local/bin")
+	if result != "/usr/local/bin" {
+		t.Errorf("NormalizeForCompare = %q; want %q", result, "/usr/local/bin")
+	}
+}
+
+func TestNormalizeForCompare_Lowercases(t *testing.T) {
+	result := NormalizeForCompare("/Foo/Bar/BAZ.go")
+	if result != "/foo/bar/baz.go" {
+		t.Errorf("NormalizeForCompare = %q; want %q", result, "/foo/bar/baz.go")
+	}
+}

--- a/internal/tools/analysis/index_search.go
+++ b/internal/tools/analysis/index_search.go
@@ -96,8 +96,8 @@ func (idx *indexer) SearchSymbols(ctx context.Context, path string, query string
 }
 
 func (idx *indexer) isInSearchPath(targetPath, filePath string) bool {
-	targetPath = filepathutil.NormalizeKey(targetPath)
-	filePath = filepathutil.NormalizeKey(filePath)
+	targetPath = filepathutil.NormalizeForCompare(targetPath)
+	filePath = filepathutil.NormalizeForCompare(filePath)
 
 	if targetPath == filePath {
 		return true

